Basic Usage
===========

TRAIN workflow steps:
    * Download of reference files and Input checking
    * Quality checking
    * Filtering of RNA-seq data
    * Alignment on reference assembly file (genome or transcriptome)
    * Read counts
    * Statistics
    * GSEA enrichment analysis

To run this workflow on a cluster, please refer to the `Snakemake documentation <https://snakemake.readthedocs.io/en/stable/executing/cluster.html>`_.

The following Snakemake command are used to test or execute each step in the TRAIN workflow.

We use the option for Conda environments: ["--use-conda", "--conda-frontend mamba"] to automatically create the required environment for each rule. Here, Mamba is used to speed up the installation.

The path to the configuration file is defined after the ["--configfile"] option. Have a look at the `configuration <https://translatome.github.io/TRAIN/configuration.html>`_ vignette.


Rule's documentation
--------------------
Here, this run shows the documentation associated with the rules using the "-l" option:

.. code-block:: bash

    $ snakemake -s translatome/workflows/00_download_references.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -l


Snakefile dry-run mode
----------------------
The dry run option allows you to see the scheduling plan including the assigned priorities before the execution of pipeline. Furthermore, this is useful to test if the workflow is defined properly and to estimate the amount of needed computation. This option is specified here with "-n".

Here, this run tests the execution in dry mode:

.. code-block:: bash

    $ snakemake -s translatome/workflows/00_download_references.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k -n


Run the snakefile
-----------------
To run the snakefile:

.. code-block:: bash

    $ snakemake -s translatome/workflows/00_download_references.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k


Execution order
---------------
The TRAIN workflow allows to use different tools for quality trimming, read counting and statistics. These tools are defined in the configuration file. 
We recommand to the users to see the results of each smk process before the next step; e.g. from the quality results depends the cleaning options to select for the cleaning tools.

.. code-block:: bash

    $ snakemake -s translatome/workflows/00_download_references.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k

    $ snakemake -s translatome/workflows/01_quality_check.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k

    $ snakemake -s translatome/workflows/02_clean_data.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k

    $ snakemake -s translatome/workflows/03_mapping.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k

    $ snakemake -s translatome/workflows/04_counts.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k

    $ snakemake -s translatome/workflows/05_stats.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k

    $ snakemake -s translatome/workflows/06_enrichment.smk --use-conda --conda-frontend mamba --configfile configs/config_PE.yml -j 1 -p -r -k


Main outputs
------------
The main figures generated correspond to:
    * a MDS plot on standardised count data, a table of standardised counts is also available.

    .. figure:: _images/MDS_example.png
        :target: _images/MDS_example.svg
        :scale: 80%
        :align: center

        Figure: Multdimensional scaling plot (MDS) on normalised data.

    * a scatter plot representing the genes categorised according to their log2FoldChange in translation and transcription. A table listing the categorised genes with their respective statistics and a summary of the number of significant genes in each category are available. 

    .. figure:: _images/scatter_example.png
        :target: _images/scatter_example.svg
        :scale: 20%
        :align: center

        Figure: Explore the log2FC data for the dual comparison to see the proportion of genes involved in translation, transcription, or both.


    * a histogram summarising the number of GO terms as a function of hierarchy depth. Several enrichment analysis tables are available in conjunction with this graph. These include gProfiler output tables and the associated summary, as well as goatools output tables showing the hierarchy of GO terms generated by gProfiler and the associated summary.

    .. figure:: _images/goatools_graph_example.png
        :target: _images/goatools_graph_example.svg
        :scale: 30%
        :align: center

        Figure: The number of terms present at each depth of the hierarchy (for BP, MF and CC) for each category for a given comparison.
